pkg/controller/resourceclaim/controller.go

```diff
@@ -0,0 +1,479 @@
+
+	queue workqueue.RateLimitingInterface
+
+	// The deletedObjects cache keeps track of Pods for which we know that
+	// they have existed and have been removed. For those we can be sure
+	// that a ReservedFor entry needs to be removed.
+	deletedObjects *uidCache
+}
+
+const (
+	claimKeyPrefix = "claim:"
+	podKeyPrefix   = "pod:"
+)
+
+// NewController creates a ResourceClaim controller.
+func NewController(
+	kubeClient clientset.Interface,
+	podInformer v1informers.PodInformer,
+	claimInformer resourcev1alpha1informers.ResourceClaimInformer,
+	templateInformer resourcev1alpha1informers.ResourceClaimTemplateInformer) (*Controller, error) {
+
+	ec := &Controller{
+		kubeClient:      kubeClient,
+		podLister:       podInformer.Lister(),
+		podIndexer:      podInformer.Informer().GetIndexer(),
+		podSynced:       podInformer.Informer().HasSynced,
+		claimLister:     claimInformer.Lister(),
+		claimsSynced:    claimInformer.Informer().HasSynced,
+		templateLister:  templateInformer.Lister(),
+		templatesSynced: templateInformer.Informer().HasSynced,
+		queue:           workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "resource_claim"),
+		deletedObjects:  newUIDCache(maxUIDCacheEntries),
+	}
+
+	metrics.RegisterMetrics()
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(klog.Infof)
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
+	ec.recorder = eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "resource_claim"})
+
+	if _, err := podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			ec.enqueuePod(obj, false)
+		},
+		UpdateFunc: func(old, updated interface{}) {
+			ec.enqueuePod(updated, false)
+		},
+		DeleteFunc: func(obj interface{}) {
+			ec.enqueuePod(obj, true)
+		},
+	}); err != nil {
+		return nil, err
+	}
+	if _, err := claimInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: ec.onResourceClaimAddOrUpdate,
+		UpdateFunc: func(old, updated interface{}) {
+			ec.onResourceClaimAddOrUpdate(updated)
+		},
+		DeleteFunc: ec.onResourceClaimDelete,
+	}); err != nil {
+		return nil, err
+	}
+	if err := ec.podIndexer.AddIndexers(cache.Indexers{podResourceClaimIndex: podResourceClaimIndexFunc}); err != nil {
+		return nil, fmt.Errorf("could not initialize ResourceClaim controller: %w", err)
+	}
+
+	return ec, nil
+}
+
+func (ec *Controller) enqueuePod(obj interface{}, deleted bool) {
+	if d, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+		obj = d.Obj
+	}
+	pod, ok := obj.(*v1.Pod)
+	if !ok {
+		// Not a pod?!
+		return
+	}
+
+	if deleted {
+		ec.deletedObjects.Add(pod.UID)
+	}
+
+	if len(pod.Spec.ResourceClaims) == 0 {
+		// Nothing to do for it at all.
+		return
+	}
+
+	// Release reservations of a deleted or completed pod?
+	if deleted ||
+		pod.Status.Phase == v1.PodFailed ||
+		pod.Status.Phase == v1.PodSucceeded ||
```

Comment:
asked about the accuracy of this in sig-node slack: https://kubernetes.slack.com/archives/C0BP8PW9G/p1668010870986709

Reply:
@bobbypage could you ack?
Although I think this should leave somewhere in the pkg/api package.

Reply:
I don't fully understand why we need to look at `pod.DeletionTimestamp` here. `pod.DeletionTimestamp` will be set once the pod starts terminating (so it's set the pod is in progress of terminating, but it's not terminated yet). Is the goal here to check if pods are already terminated? In that case, we should just check if 

1. The pod is deleted
2. The pod is in terminal phase - https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/pod/util.go#L300-L303

Reply:
Because the Pod might not have been scheduled.

Reply:
> Because the Pod might not have been scheduled.

Hmm, if we want to check if the pod has not been scheduled why would we look at pod.DeletionTimestamp? Shouldn't we look at `len(spec.nodeName) == 0` then?

Reply:
`spec.nodeName == ""` alone only implies "not scheduled yet". `spec.nodeName == "" && pod.DeletionTimestamp != nil` implies "will never get scheduled", which is a relevant difference here.

Reply:
I see, makes sense. Can we convert the `pod.Status.Phase` checks to use `IsPodTerminal` - https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/pod/util.go#L300-L303?

Reply:
Yes, done. Thanks for the pointer.